hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -818,57 +817,57 @@ public boolean isEmpty() {
-                        = new HashMap<Integer, MapEntrySet>(nodeEngine.getPartitionService().getPartitionCount());
-                for (Entry entry : entries.entrySet()) {
-                    checkNotNull(entry.getKey(), NULL_KEY_IS_NOT_ALLOWED);
-                    checkNotNull(entry.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
-
-                    int partitionId = partitionService.getPartitionId(entry.getKey());
-                    if (!entryMap.containsKey(partitionId)) {
-                        entryMap.put(partitionId, new MapEntrySet());
-                    }
-                    entryMap.get(partitionId).add(
-                            new AbstractMap.SimpleImmutableEntry<Data, Data>(mapService.getMapServiceContext().toData(
-                                    entry.getKey(),
-                                    partitionStrategy),
-                                    mapService.getMapServiceContext()
-                                            .toData(entry.getValue())
-                            ));
+            List<Future> futures = new ArrayList<Future>(partitionCount);
+            MapEntrySet[] entrySetPerPartition = new MapEntrySet[partitionCount];
+
+            // first we fill entrySetPerPartition
+            for (Entry entry : entries.entrySet()) {
+                checkNotNull(entry.getKey(), NULL_KEY_IS_NOT_ALLOWED);
+                checkNotNull(entry.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
+
+                int partitionId = partitionService.getPartitionId(entry.getKey());
+                MapEntrySet entrySet = entrySetPerPartition[partitionId];
+                if (entrySet == null) {
+                    entrySet = new MapEntrySet();
+                    entrySetPerPartition[partitionId] = entrySet;
                 }
 
-                for (final Map.Entry<Integer, MapEntrySet> entry : entryMap.entrySet()) {
-                    final Integer partitionId = entry.getKey();
-                    final PutAllOperation op = new PutAllOperation(name, entry.getValue());
-                    op.setPartitionId(partitionId);
-                    futures.add(operationService.invokeOnPartition(SERVICE_NAME, op, partitionId));
-                }
+                Data keyData = mapServiceContext.toData(entry.getKey(), partitionStrategy);
+                Data valueData = mapServiceContext.toData(entry.getValue());
+                entrySet.add(new AbstractMap.SimpleImmutableEntry<Data, Data>(keyData, valueData));
+            }
 
-                for (Future future : futures) {
-                    future.get();
+            // then we invoke the operations
+            for (int partitionId = 0; partitionId < entrySetPerPartition.length; partitionId++) {
+                MapEntrySet entrySet = entrySetPerPartition[partitionId];
+                if (entrySet != null) {
+                    // If there is a single entry, we could make use of a PutOperation since that is a bit cheaper
```

Comment:
We probably want to avoid a PutOperation given that we want to avoid getting back the old value that maps to the key. See #6083.